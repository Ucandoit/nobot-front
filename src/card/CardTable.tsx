import { Avatar } from '@material-ui/core';
import { makeStyles } from '@material-ui/core/styles';
import Table from '@material-ui/core/Table';
import TableBody from '@material-ui/core/TableBody';
import TableCell from '@material-ui/core/TableCell';
import TableContainer from '@material-ui/core/TableContainer';
import TableHead from '@material-ui/core/TableHead';
import TableRow from '@material-ui/core/TableRow';
import TableSortLabel from '@material-ui/core/TableSortLabel';
import React, { useCallback } from 'react';
import { Order } from '../commons';
import { Card, utils } from '../helpers';

const useStyles = makeStyles(theme => ({
  root: {
    width: '100%'
  },
  paper: {
    width: '100%',
    marginBottom: theme.spacing(2)
  },
  table: {
    minWidth: 750
  },
  header: {
    backgroundColor: '#c5cae9'
  },
  visuallyHidden: {
    border: 0,
    clip: 'rect(0 0 0 0)',
    height: 1,
    margin: -1,
    overflow: 'hidden',
    padding: 0,
    position: 'absolute',
    top: 20,
    width: 1
  },
  smallCell: {
    maxWidth: '4rem'
  },
  faceImg: {
    height: '2rem',
    width: '2rem'
  }
}));

interface CardTableHeader {
  label: string;
  property: keyof Card;
  sortable: boolean;
  colWidth: 'small' | 'normal' | 'large';
}

const headers: CardTableHeader[] = [];

const addHeader = (
  label: string,
  property: keyof Card,
  sortable = true,
  colWidth: 'small' | 'normal' | 'large' = 'small'
) => {
  headers.push({
    label,
    property,
    sortable,
    colWidth
  });
};

addHeader('No.', 'number');
addHeader('Img', 'faceUrl', false);
addHeader('Name', 'name', true, 'normal');
addHeader('Rarity', 'rarity');
addHeader('Property', 'property');
addHeader('Cost', 'cost');
addHeader('Military', 'military');
addHeader('Job', 'job');
addHeader('Atk', 'finalAtk');
addHeader('Def', 'finalDef');
addHeader('Spd', 'finalSpd');
addHeader('Vir', 'finalVir');
addHeader('Stg', 'finalStg');

interface CardTableProps {
  cards: Card[];
  sort: keyof Card;
  order: Order;
  changeSort: (sort: keyof Card) => void;
}

const CardTable = ({ cards, sort, order, changeSort }: CardTableProps) => {
  const classes = useStyles();

  const onTableHeaderClick = useCallback(
    (property: keyof Card) => () => {
      changeSort(property);
    },
    [changeSort]
  );

  return (
    <TableContainer>
      <Table className={classes.table} aria-labelledby="tableTitle" size="small" aria-label="card table">
        <TableHead className={classes.header}>
          <TableRow>
            {headers.map(({ label, property, sortable, colWidth }) => (
              <TableCell
                key={property}
                className={colWidth === 'small' ? classes.smallCell : ''}
                sortDirection={sort === property ? order : false}
              >
                {sortable ? (
                  <TableSortLabel
                    active={sort === property}
                    direction={sort === property ? order : 'asc'}
                    onClick={onTableHeaderClick(property)}
                  >
                    {label}
                  </TableSortLabel>
                ) : (
                  label
                )}
              </TableCell>
            ))}
          </TableRow>
        </TableHead>
        <TableBody>
          {cards.map(card => (
            <TableRow hover key={card.id}>
              <TableCell>{card.number}</TableCell>
              <TableCell>
                <Avatar src={card.faceUrl} variant="square" className={classes.faceImg} />
              </TableCell>
              <TableCell>{card.name}</TableCell>
              <TableCell>
                {card.rarity}
                {utils.getStar(card.star)}
              </TableCell>
              <TableCell>{card.property}</TableCell>
              <TableCell>{card.cost}</TableCell>
              <TableCell>{card.military}</TableCell>
              <TableCell>{card.job}</TableCell>
              <TableCell>{card.finalAtk}</TableCell>
              <TableCell>{card.finalDef}</TableCell>
              <TableCell>{card.finalSpd}</TableCell>
              <TableCell>{card.finalVir}</TableCell>
              <TableCell>{card.finalStg}</TableCell>
            </TableRow>
          ))}
        </TableBody>
      </Table>
    </TableContainer>
  );
};

export default CardTable;
