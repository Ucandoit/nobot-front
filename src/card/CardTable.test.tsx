import { TableHead, TableSortLabel } from '@material-ui/core';
import { shallow, ShallowWrapper } from 'enzyme';
import React from 'react';
import { cards } from './cards.mock';
import CardTable from './CardTable';

const changeSort = jest.fn();

describe('CardTable', () => {
  let wrapper: ShallowWrapper;
  beforeEach(() => {
    wrapper = shallow(<CardTable cards={cards} sort="number" order="asc" changeSort={changeSort} />);
  });

  it('should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should call changeSort function on column header click', () => {
    wrapper
      .find(TableHead)
      .find(TableSortLabel)
      .at(1)
      .simulate('click');
    expect(changeSort).toHaveBeenCalledTimes(1);
    expect(changeSort).toHaveBeenCalledWith('name');
  });
});
