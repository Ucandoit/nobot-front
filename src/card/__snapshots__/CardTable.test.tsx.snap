// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardTable should render correctly 1`] = `
<WithStyles(ForwardRef(TableContainer))>
  <WithStyles(ForwardRef(Table))
    aria-label="card table"
    aria-labelledby="tableTitle"
    className="makeStyles-table-3"
    size="small"
  >
    <WithStyles(ForwardRef(TableHead))
      className="makeStyles-header-4"
    >
      <WithStyles(ForwardRef(TableRow))>
        <WithStyles(ForwardRef(TableCell))
          className="makeStyles-smallCell-6"
          key="number"
          sortDirection="asc"
        >
          <WithStyles(ForwardRef(TableSortLabel))
            active={true}
            direction="asc"
            onClick={[Function]}
          >
            No.
          </WithStyles(ForwardRef(TableSortLabel))>
        </WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(TableCell))
          className="makeStyles-smallCell-6"
          key="faceUrl"
          sortDirection={false}
        >
          Img
        </WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(TableCell))
          className=""
          key="name"
          sortDirection={false}
        >
          <WithStyles(ForwardRef(TableSortLabel))
            active={false}
            direction="asc"
            onClick={[Function]}
          >
            Name
          </WithStyles(ForwardRef(TableSortLabel))>
        </WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(TableCell))
          className="makeStyles-smallCell-6"
          key="rarity"
          sortDirection={false}
        >
          <WithStyles(ForwardRef(TableSortLabel))
            active={false}
            direction="asc"
            onClick={[Function]}
          >
            Rarity
          </WithStyles(ForwardRef(TableSortLabel))>
        </WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(TableCell))
          className="makeStyles-smallCell-6"
          key="property"
          sortDirection={false}
        >
          <WithStyles(ForwardRef(TableSortLabel))
            active={false}
            direction="asc"
            onClick={[Function]}
          >
            Property
          </WithStyles(ForwardRef(TableSortLabel))>
        </WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(TableCell))
          className="makeStyles-smallCell-6"
          key="cost"
          sortDirection={false}
        >
          <WithStyles(ForwardRef(TableSortLabel))
            active={false}
            direction="asc"
            onClick={[Function]}
          >
            Cost
          </WithStyles(ForwardRef(TableSortLabel))>
        </WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(TableCell))
          className="makeStyles-smallCell-6"
          key="military"
          sortDirection={false}
        >
          <WithStyles(ForwardRef(TableSortLabel))
            active={false}
            direction="asc"
            onClick={[Function]}
          >
            Military
          </WithStyles(ForwardRef(TableSortLabel))>
        </WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(TableCell))
          className="makeStyles-smallCell-6"
          key="job"
          sortDirection={false}
        >
          <WithStyles(ForwardRef(TableSortLabel))
            active={false}
            direction="asc"
            onClick={[Function]}
          >
            Job
          </WithStyles(ForwardRef(TableSortLabel))>
        </WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(TableCell))
          className="makeStyles-smallCell-6"
          key="finalAtk"
          sortDirection={false}
        >
          <WithStyles(ForwardRef(TableSortLabel))
            active={false}
            direction="asc"
            onClick={[Function]}
          >
            Atk
          </WithStyles(ForwardRef(TableSortLabel))>
        </WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(TableCell))
          className="makeStyles-smallCell-6"
          key="finalDef"
          sortDirection={false}
        >
          <WithStyles(ForwardRef(TableSortLabel))
            active={false}
            direction="asc"
            onClick={[Function]}
          >
            Def
          </WithStyles(ForwardRef(TableSortLabel))>
        </WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(TableCell))
          className="makeStyles-smallCell-6"
          key="finalSpd"
          sortDirection={false}
        >
          <WithStyles(ForwardRef(TableSortLabel))
            active={false}
            direction="asc"
            onClick={[Function]}
          >
            Spd
          </WithStyles(ForwardRef(TableSortLabel))>
        </WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(TableCell))
          className="makeStyles-smallCell-6"
          key="finalVir"
          sortDirection={false}
        >
          <WithStyles(ForwardRef(TableSortLabel))
            active={false}
            direction="asc"
            onClick={[Function]}
          >
            Vir
          </WithStyles(ForwardRef(TableSortLabel))>
        </WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(TableCell))
          className="makeStyles-smallCell-6"
          key="finalStg"
          sortDirection={false}
        >
          <WithStyles(ForwardRef(TableSortLabel))
            active={false}
            direction="asc"
            onClick={[Function]}
          >
            Stg
          </WithStyles(ForwardRef(TableSortLabel))>
        </WithStyles(ForwardRef(TableCell))>
      </WithStyles(ForwardRef(TableRow))>
    </WithStyles(ForwardRef(TableHead))>
    <WithStyles(ForwardRef(TableBody))>
      <WithStyles(ForwardRef(TableRow))
        hover={true}
        key="3"
      >
        <WithStyles(ForwardRef(TableCell))>
          1001
        </WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(TableCell))>
          <WithStyles(ForwardRef(Avatar))
            className="makeStyles-faceImg-7"
            src="http://210.140.157.168/img/card/still/character_still_00034_2067324248.png"
            variant="square"
          />
        </WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(TableCell))>
          上杉ニャンしん
        </WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(TableCell))>
          極
        </WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(TableCell))>
          火
        </WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(TableCell))>
          3
        </WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(TableCell))>
          騎馬
        </WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(TableCell))>
          武士
        </WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(TableCell))>
          197
        </WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(TableCell))>
          140
        </WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(TableCell))>
          164
        </WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(TableCell))>
          142
        </WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(TableCell))>
          166
        </WithStyles(ForwardRef(TableCell))>
      </WithStyles(ForwardRef(TableRow))>
      <WithStyles(ForwardRef(TableRow))
        hover={true}
        key="4"
      >
        <WithStyles(ForwardRef(TableCell))>
          1002
        </WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(TableCell))>
          <WithStyles(ForwardRef(Avatar))
            className="makeStyles-faceImg-7"
            src="http://210.140.157.168/img/card/still/character_still_00040_1306553348.png"
            variant="square"
          />
        </WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(TableCell))>
          おミーちの方
        </WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(TableCell))>
          極
        </WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(TableCell))>
          水
        </WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(TableCell))>
          3
        </WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(TableCell))>
          足軽
        </WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(TableCell))>
          姫
        </WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(TableCell))>
          144
        </WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(TableCell))>
          164
        </WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(TableCell))>
          162
        </WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(TableCell))>
          198
        </WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(TableCell))>
          143
        </WithStyles(ForwardRef(TableCell))>
      </WithStyles(ForwardRef(TableRow))>
      <WithStyles(ForwardRef(TableRow))
        hover={true}
        key="5"
      >
        <WithStyles(ForwardRef(TableCell))>
          1003
        </WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(TableCell))>
          <WithStyles(ForwardRef(Avatar))
            className="makeStyles-faceImg-7"
            src="http://210.140.157.168/img/card/still/character_still_00050_2564820845.png"
            variant="square"
          />
        </WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(TableCell))>
          織田のぶニャが
        </WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(TableCell))>
          極
        </WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(TableCell))>
          火
        </WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(TableCell))>
          3
        </WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(TableCell))>
          鉄砲
        </WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(TableCell))>
          武士
        </WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(TableCell))>
          186
        </WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(TableCell))>
          166
        </WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(TableCell))>
          162
        </WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(TableCell))>
          137
        </WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(TableCell))>
          158
        </WithStyles(ForwardRef(TableCell))>
      </WithStyles(ForwardRef(TableRow))>
      <WithStyles(ForwardRef(TableRow))
        hover={true}
        key="7"
      >
        <WithStyles(ForwardRef(TableCell))>
          1004
        </WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(TableCell))>
          <WithStyles(ForwardRef(Avatar))
            className="makeStyles-faceImg-7"
            src="http://210.140.157.168/img/card/still/character_still_00100_142767575.png"
            variant="square"
          />
        </WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(TableCell))>
          さニャだ幸村
        </WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(TableCell))>
          極
        </WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(TableCell))>
          火
        </WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(TableCell))>
          2.5
        </WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(TableCell))>
          騎馬
        </WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(TableCell))>
          武士
        </WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(TableCell))>
          184
        </WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(TableCell))>
          115
        </WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(TableCell))>
          117
        </WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(TableCell))>
          135
        </WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(TableCell))>
          130
        </WithStyles(ForwardRef(TableCell))>
      </WithStyles(ForwardRef(TableRow))>
      <WithStyles(ForwardRef(TableRow))
        hover={true}
        key="8"
      >
        <WithStyles(ForwardRef(TableCell))>
          1005
        </WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(TableCell))>
          <WithStyles(ForwardRef(Avatar))
            className="makeStyles-faceImg-7"
            src="http://210.140.157.168/img/card/still/character_still_00109_1467951905.png"
            variant="square"
          />
        </WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(TableCell))>
          シャムづ義弘
        </WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(TableCell))>
          極
        </WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(TableCell))>
          地
        </WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(TableCell))>
          3
        </WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(TableCell))>
          騎馬
        </WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(TableCell))>
          武士
        </WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(TableCell))>
          180
        </WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(TableCell))>
          193
        </WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(TableCell))>
          149
        </WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(TableCell))>
          131
        </WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(TableCell))>
          157
        </WithStyles(ForwardRef(TableCell))>
      </WithStyles(ForwardRef(TableRow))>
    </WithStyles(ForwardRef(TableBody))>
  </WithStyles(ForwardRef(Table))>
</WithStyles(ForwardRef(TableContainer))>
`;
